# Titanic Survival

#### Grading:


- Code: 90 pts
- Markdown Documentation: 10 pts


We are going to study the survival rate of passengers on titanic and what variables affected survival.

Load the dataset in `titanic.xls`. It contains data on all the passengers that travelled on the Titanic.

In [1]:
from IPython.core.display import HTML
HTML(filename='../data/titanic.html')

In [2]:
import pandas as pd
# you would need xlrd - pip install xlrd
t_file = pd.ExcelFile('../data/titanic.xls')
t_df = t_file.parse("titanic", header=None)
t_df.head()

0         1                                     2       3       4   \
0  pclass  survived                                  name     sex     age   
1       1         1         Allen, Miss. Elisabeth Walton  female      29   
2       1         1        Allison, Master. Hudson Trevor    male  0.9167   
3       1         0          Allison, Miss. Helen Loraine  female       2   
4       1         0  Allison, Mr. Hudson Joshua Creighton    male      30   

      5      6       7        8        9         10    11    12  \
0  sibsp  parch  ticket     fare    cabin  embarked  boat  body   
1      0      0   24160  211.338       B5         S     2   NaN   
2      1      2  113781   151.55  C22 C26         S    11   NaN   
3      1      2  113781   151.55  C22 C26         S   NaN   NaN   
4      1      2  113781   151.55  C22 C26         S   NaN   135   

                                13  
0                        home.dest  
1                     St Louis, MO  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

### Women and children first?

*** 1. Use the `groupby` method to calculate the proportion of passengers that survived by sex. (25 pts)*** 

In [3]:
t_df.columns = t_df.iloc[0]
t_df.drop(t_df.index[0], inplace=True)
t_df.head()

0 pclass survived                                             name     sex  \
1      1        1                    Allen, Miss. Elisabeth Walton  female   
2      1        1                   Allison, Master. Hudson Trevor    male   
3      1        0                     Allison, Miss. Helen Loraine  female   
4      1        0             Allison, Mr. Hudson Joshua Creighton    male   
5      1        0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

0     age sibsp parch  ticket     fare    cabin embarked boat body  \
1      29     0     0   24160  211.338       B5        S    2  NaN   
2  0.9167     1     2  113781   151.55  C22 C26        S   11  NaN   
3       2     1     2  113781   151.55  C22 C26        S  NaN  NaN   
4      30     1     2  113781   151.55  C22 C26        S  NaN  135   
5      25     1     2  113781   151.55  C22 C26        S  NaN  NaN   

0                        home.dest  
1                     St Louis, MO  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON  
5  Montreal, PQ / Chesterville, ON

Shows How many total passengers there are for each gender.

In [4]:
survival = t_df.copy()
survival.groupby('sex')['survived'].count()

sex
female    466
male      843
Name: survived, dtype: int64

Shows the number of how many people of each gender had died and survived.

In [5]:
sex_ratio = t_df.copy()
sex_ratio['proportion'] = 1
sex_ratio = sex_ratio.groupby(['sex', 'survived']).agg({'proportion': 'sum'})
print(sex_ratio)

                 proportion
sex    survived            
female 0                127
       1                339
male   0                682
       1                161


Calulates and prints the proportion of each gender that survived and died.

In [6]:
survivedProportion = sex_ratio.groupby(level=0).apply(lambda x:100 * x/x.sum())
print(survivedProportion)

                 proportion
sex    survived            
female 0          27.253219
       1          72.746781
male   0          80.901542
       1          19.098458


## This calculates and prints the percentage of each gender that had survived. Using a try and except block.

In [7]:
part1= t_df.copy()
sex_survive = part1.groupby(['sex','survived'])
sex = ['female', 'male']

for gender in sex:
    try:
        death = len(sex_survive.get_group((gender, 0)))
        survived = len(sex_survive.get_group((gender, 1)))
        print('\n The percentage of survival for', gender,': ', (survived/(death+survived)*100))
    except:
        continue


 The percentage of survival for female :  72.74678111587983

 The percentage of survival for male :  19.098457888493474


#### 72.75% of the woman had survived.

#### 19.09% ~ 19.1% of the men survived.

*** 2. Calculate the same proportion, but by class and sex. (25 pts)*** 

Shows the total number of each gender in each class.

In [8]:
survival_class_ratio =t_df.copy()
survival_class_ratio.groupby(['pclass','sex'])['survived'].count()

pclass  sex   
1       female    144
        male      179
2       female    106
        male      171
3       female    216
        male      493
Name: survived, dtype: int64

Shows the number of people in each gender of each class that had survived and died.

In [9]:
pclass = t_df.copy()
pclass['proportion'] = 1
pclass = pclass.groupby(['pclass','sex', 'survived']).agg({'proportion': 'sum'})
print(pclass)

                        proportion
pclass sex    survived            
1      female 0                  5
              1                139
       male   0                118
              1                 61
2      female 0                 12
              1                 94
       male   0                146
              1                 25
3      female 0                110
              1                106
       male   0                418
              1                 75


This only calculates and prints the pclass proportions. 

In [10]:
pclassProportion = pclass.groupby(level=0).apply(lambda x:100 * x/x.sum())
print(pclassProportion)

                        proportion
pclass sex    survived            
1      female 0           1.547988
              1          43.034056
       male   0          36.532508
              1          18.885449
2      female 0           4.332130
              1          33.935018
       male   0          52.707581
              1           9.025271
3      female 0          15.514810
              1          14.950635
       male   0          58.956276
              1          10.578279


Gathers the totals of each gender in each class.

In [11]:
male = t_df[t_df['sex'] == "male"]
male_class_1 = male[male['pclass'] == 1]
male_class_2 = male[male['pclass'] == 2]
male_class_3 = male[male['pclass'] == 3]
female = t_df[t_df['sex'] == "female"]
female_class_1 = female[female['pclass'] == 1]
female_class_2 = female[female['pclass'] == 2]
female_class_3 = female[female['pclass'] == 3]

Calculates the survival rate of males in each class.

In [12]:
61/len(male_class_1) * 100, 25/len(male_class_2) * 100, 75/len(male_class_3) * 100

(34.07821229050279, 14.619883040935672, 15.212981744421908)

Calculates the survival rate of females in each class.

In [13]:
139/len(female_class_1) * 100, 94/len(female_class_2) * 100, 106/len(female_class_3) * 100

(96.52777777777779, 88.67924528301887, 49.074074074074076)

#### Calculates and prints the survival percentage of each gender in each glass using try and catch block.

In [14]:
part2= t_df.copy()
pclass_sex = part2.groupby(['pclass','sex','survived'])
pclass = [1,2,3]

for gender in sex:
    print('\n The percentage of survival for', gender,': ')
    for number in pclass:
        try:
            death = len(pclass_sex.get_group((number, gender, 0)))
            survived = len(pclass_sex.get_group((number, gender, 1)))
            print('\n pclass', number,': ', (survived/(death+survived)*100))
        except:
            continue


 The percentage of survival for female : 

 pclass 1 :  96.52777777777779

 pclass 2 :  88.67924528301887

 pclass 3 :  49.074074074074076

 The percentage of survival for male : 

 pclass 1 :  34.07821229050279

 pclass 2 :  14.619883040935672

 pclass 3 :  15.212981744421908


*** 3. Create age categories: children (under 14 years), adolescents (14-20), adult (21-64), and senior(65+), and calculate survival proportions by age category, class and sex. ( pts)***

#### This makes a new table and lables each age category accordingly.

In [15]:
new_t_df = t_df.copy()

new_t_df["Age_Group"] = ""
for x in range(1, len(new_t_df)+1):
    if (new_t_df.age[x] < 14.0): 
        new_t_df.Age_Group[x] = "children"
    elif (new_t_df.age[x] > 13.0) and (new_t_df.age[x] < 21.0):
        new_t_df.Age_Group[x] = "adolescents"
    elif (new_t_df.age[x] > 20.0) and (new_t_df.age[x] < 65.0):
        new_t_df.Age_Group[x] = "adult"
    elif (new_t_df.age[x] > 64.0):
        new_t_df.Age_Group[x] = "senior"
    else:
        new_t_df.Age_Group[x] = "NaN"
new_t_df.head(20)        

0  pclass survived                                               name     sex  \
1       1        1                      Allen, Miss. Elisabeth Walton  female   
2       1        1                     Allison, Master. Hudson Trevor    male   
3       1        0                       Allison, Miss. Helen Loraine  female   
4       1        0               Allison, Mr. Hudson Joshua Creighton    male   
5       1        0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
6       1        1                                Anderson, Mr. Harry    male   
7       1        1                  Andrews, Miss. Kornelia Theodosia  female   
8       1        0                             Andrews, Mr. Thomas Jr    male   
9       1        1      Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
10      1        0                            Artagaveytia, Mr. Ramon    male   
11      1        0                             Astor, Col. John Jacob    male   
12      1        1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female   
13      1        1                      Aubart, Mme. Leontine Pauline  female   
14      1        1                       Barber, Miss. Ellen "Nellie"  female   
15      1        1               Barkworth, Mr. Algernon Henry Wilson    male   
16      1        0                                Baumann, Mr. John D    male   
17      1        0                           Baxter, Mr. Quigg Edmond    male   
18      1        1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female   
19      1        1                              Bazzani, Miss. Albina  female   
20      1        0                               Beattie, Mr. Thomson    male   

0      age sibsp parch    ticket     fare    cabin embarked boat body  \
1       29     0     0     24160  211.338       B5        S    2  NaN   
2   0.9167     1     2    113781   151.55  C22 C26        S   11  NaN   
3        2     1     2    113781   151.55  C22 C26        S  NaN  NaN   
4       30     1     2    113781   151.55  C22 C26        S  NaN  135   
5       25     1     2    113781   151.55  C22 C26        S  NaN  NaN   
6       48     0     0     19952    26.55      E12        S    3  NaN   
7       63     1     0     13502  77.9583       D7        S   10  NaN   
8       39     0     0    112050        0      A36        S  NaN  NaN   
9       53     2     0     11769  51.4792     C101        S    D  NaN   
10      71     0     0  PC 17609  49.5042      NaN        C  NaN   22   
11      47     1     0  PC 17757  227.525  C62 C64        C  NaN  124   
12      18     1     0  PC 17757  227.525  C62 C64        C    4  NaN   
13      24     0     0  PC 17477     69.3      B35        C    9  NaN   
14      26     0     0     19877    78.85      NaN        S    6  NaN   
15      80     0     0     27042       30      A23        S    B  NaN   
16     NaN     0     0  PC 17318   25.925      NaN        S  NaN  NaN   
17      24     0     1  PC 17558  247.521  B58 B60        C  NaN  NaN   
18      50     0     1  PC 17558  247.521  B58 B60        C    6  NaN   
19      32     0     0     11813  76.2917      D15        C    8  NaN   
20      36     0     0     13050  75.2417       C6        C    A  NaN   

0                         home.dest    Age_Group  
1                      St Louis, MO        adult  
2   Montreal, PQ / Chesterville, ON     children  
3   Montreal, PQ / Chesterville, ON     children  
4   Montreal, PQ / Chesterville, ON        adult  
5   Montreal, PQ / Chesterville, ON        adult  
6                      New York, NY        adult  
7                        Hudson, NY        adult  
8                       Belfast, NI        adult  
9               Bayside, Queens, NY        adult  
10              Montevideo, Uruguay       senior  
11                     New York, NY        adult  
12                     New York, NY  adolescents  
13                    Paris, France        adult  
14                              NaN        adult  

This is the proportions of how many survived and died in each age, sex, and class category.

In [16]:
new_t_df['proportion'] = 1
age_group = new_t_df.groupby(['Age_Group','pclass','sex', 'survived']).agg({'proportion': 'sum'})
print(age_group)

                                    proportion
Age_Group   pclass sex    survived            
NaN         1      female 1                 11
                   male   0                 20
                          1                  8
            2      female 0                  1
                          1                  2
                   male   0                 11
                          1                  2
            3      female 0                 30
                          1                 34
                   male   0                128
                          1                 16
adolescents 1      female 1                 15
                   male   0                  4
                          1                  1
            2      female 0                  1
                          1                 12
                   male   0                 15
                          1                  2
            3      female 0                 16
             

##### This calculates and prints the percentage of survival rate based on age, class, and sex based off the new table with age categories using a try catch block.

In [17]:
part3 = new_t_df.copy()
def ages (Age_Group):
    if Age_Group == "children": return 0
    elif Age_Group == "adolescents": return 1
    elif Age_Group == "adult": return 2
    elif Age_Group == "senior": return 3
    else: return 4
    
part3['Age_Group'] = part3['Age_Group'].apply(lambda x: ages(x))

age_pclass_sex = part3.groupby(['Age_Group','pclass','sex','survived'])
age = [0, 1, 2, 3, 4]

for gender in sex:
    print('\n The percentage of survival for', gender,': ')
    for number in pclass:
        print('\n pclass', number)
        for age_group in age:
            try:
                death = len(age_pclass_sex.get_group((age_group, number, gender, 0)))
                survived = len(age_pclass_sex.get_group((age_group, number, gender, 1)))
                print(' age group', age_group,': ', (survived/(death+survived)*100))
            except:
                continue


 The percentage of survival for female : 

 pclass 1
 age group 2 :  96.55172413793103

 pclass 2
 age group 1 :  92.3076923076923
 age group 2 :  86.8421052631579
 age group 4 :  66.66666666666666

 pclass 3
 age group 0 :  48.38709677419355
 age group 1 :  54.285714285714285
 age group 2 :  44.18604651162791
 age group 4 :  53.125

 The percentage of survival for male : 

 pclass 1
 age group 1 :  20.0
 age group 2 :  34.32835820895522
 age group 3 :  14.285714285714285
 age group 4 :  28.57142857142857

 pclass 2
 age group 1 :  11.76470588235294
 age group 2 :  7.8125
 age group 4 :  15.384615384615385

 pclass 3
 age group 0 :  32.432432432432435
 age group 1 :  12.307692307692308
 age group 2 :  15.983606557377051
 age group 4 :  11.11111111111111
